sympy/sets/handlers/functions.py

```diff
@@ -148,11 +148,15 @@ def _set_function(f, x): # noqa:F811
 @dispatch(FunctionUnion, Range)  # type: ignore # noqa:F811
 def _set_function(f, self): # noqa:F811
     from sympy.core.function import expand_mul
-    if not self:
+    try:
+        n = self.size
+    except ValueError:
+        n = None
```

Comment:
I would prefer an explicit check rather than catching an exception.

Reply:
@oscarbenjamin i mean like I've tried an explicit check ie something like
 if not self.size.is_Integer:
    return
but it kinda seems to return the initial error of the problem ie the ValueError('invalid method for symbolic range'),
what could be a possible solution to this?

Reply:
Perhaps Range could be modified to return None instead of an error -- but I'm not sure what the implications of that would be. Capturing the ValueError for a single line of code that queries a property that raises a ValueError if the size is unknown doesn't bother me so much, however. (But I'm the one that suggested it.)

Reply:
There should be a way to get the symbolic size. It could be a different attribute like `r.length` or something but the size of `Range(n)` is `n` and it should be possible to get that in some way without an exception.

Reply:
Only a few symbolic assumptions could actually give you a size (like even+prime=2). And it is worse with `Range(x,y,z)`. Symbolic ranges are largely for display purposes only; is it worth the work to return a Piecewise that would be true for symbolic arguments? And that should only be returned if all args are declared as integers. It's not a trivial matter; I doubt it is of any significant use; I would prefer to just let the ValueError signal that it's not available (as here) and let the code work around it. Would returning None be better than raising an error? Or would `NotImplemented('on purpose')` be better?

Reply:
If we are going to have a Range with symbolic start/end then its size can be symbolic as well. I can see the value in having an attribute that gives the size when it is known as an explicit integer and raises otherwise (as `.size` does). There also is a need for an attribute that simply gives the symbolic size as an expression e.g. `Range(n).length -> n` even if `n` is a symbol or some other expression.

Many builtin operations can raise ValueError so ValueError should not be explicitly raised and caught within the SymPy codebase because it can mask bugs. If there is ever a need to catch the exception then a different exception class should be raised in the first place.

It is more robust not to catch exceptions at all though. There should be an interface that makes it possible to check whether an operation should succeed without needing to catch an exception: then any exception that does get raised can bubble up and can rightly be considered a bug.

Reply:
This is a test of the Piecewise logic for the step size (assuming 0 for cases that would raise an error):
```python
from itertools import combinations, permutations
for i in combinations(range(-3,4),3):
    for i in permutations(i):
        start,stop,step=i
        if all([int(start)==start,int(stop)==stop,int(step)==step,(stop-start)*step>0]):
            n = int((stop - start)/step) + (1 if (stop-start)%step else 0)
        else:
            n=0
        try:
            m=len(range(*i))
        except:
            m=0
        assert n == m
```
And here is the Piecewise which checks that all args are integers or at most 1 of them is infinite, that they are compatible with the direction of the step:
```python
Piecewise((floor((stop - start)/step) + Piecewise((1,Ne(Mod(stop-start,step),0)),(0,True)),And(Eq(floor(start),start),Eq(floor(stop),stop),Eq(floor(step),step),(stop - start)*step>0,Ne(step+1,step),Or(Ne(start+1,start),Ne(stop+1,stop)))),(0,True))
```

Reply:
Okay, that's very complicated. That makes me wonder if symbolic range was a good idea. We mainly use it because we want to represent sets of integers but it would be better to have something simpler for that e.g.: https://github.com/sympy/sympy/issues/17856#issuecomment-552229762

Reply:
We supported it for the purpose of code generation. We could also compute the size for something like `Range(i, i + k*j, k)` as `j` but I think for simplicity we just steered clear of trying to support it without a well defined need.